### DISCUSSION TITLE
How to determine what's wrong with a file when UVtools cannot open it with crash

### DISCUSSION THREAD

**Discussion by Megidd** [sn4k3/UVtools#216] [Q&A] [answered]
I try to open this faulty CBDDLP file. The UVtools crashes and cannot open it. Is there any tool to figure out why UVtools crashes, i.e. to find out what's wrong with the file?

[teapot_2021-06-09 (3).zip](https://github.com/sn4k3/UVtools/files/6623369/teapot_2021-06-09.3.zip)

The crash happens at random times:

* When loading thumbnails
* When loading layers
* ...

Some screenshots right before the crash:

![image](https://user-images.githubusercontent.com/17475482/121338136-7695ab00-c932-11eb-9cfd-f00a939e752e.png)

![image](https://user-images.githubusercontent.com/17475482/121338166-7d242280-c932-11eb-8f5c-84da5546f4a0.png)

**Comment by sn4k3** [comment, marked as answer]
Are you on a Linux OS?
1. Go to UVtools -> Help -> Open settings folder
1. Open the "errors.log"

**EDIT:** Ok i found the problem. That file is very old, is from **Version: 1** of the format which UVtools do not fully known. It's only ready to deal with v2 and v3. That file must have years... Even PhotonFileValidator can't read it. Somehow is corrupted. You can try run on printer without vat to see if it runs.

**Reply by Megidd** [reply]
Thanks! I'm on Windows 10. For some reason, the settings folder is empty.

**Reply by sn4k3** [reply]
If empty then there are no produced errors yet

**Reply by Megidd** [reply]
Thanks! 😃 

* Can you try this file too? Is its version old too?
* Can I run UVtools by the command line to get any possible error log?

[teapot_2021-06-09 (7).zip](https://github.com/sn4k3/UVtools/files/6624784/teapot_2021-06-09.7.zip)

**Reply by sn4k3** [reply]
Ok that file opens and it's version 1. Meaning the first you sent is somehow corrupted. It had problems on preview, and exposure time was crazy high numbers
![image](https://user-images.githubusercontent.com/113281/121380507-d3965e80-c93c-11eb-8a92-3429c9b38eaf.png)

**Reply by Megidd** [reply]
Great. The first file and the second one both are created by almost the same process. For some reason the first one is corrupt. I plan to run UVtools by attaching a debugger to it to see why it crashes. Are there any instructions to help me attach a debugger to UVtools? Thanks!

**Reply by sn4k3** [reply]
You need visual studio for that, you can run it as DEBUG with source code, or attach the process and load the pdb. Most of the times the UI will return a error, but in a very bad case like that one which attempt to access to private memory process will be killed, that's why you see no error.

**Reply by sn4k3** [reply]
btw what slicer are you using that creates a 1161x394 thumbnails?

**Reply by Megidd** [reply]
I'm not using any slicer! 😄 

Actually, I'm modifying [this repository](https://github.com/Andoryuuta/photon) to generate CDBDLP/Photon file format from a collection of PNG files using Golang. 

I'm using a single PNG file for both preview image and thumbnail image for now. In the future, having debugged the modified repository, I will use a smaller PNG file for the thumbnail.

* The 2nd attached file i.e. the good one was a re-export of a correct Photon file by [the original repository](https://github.com/Andoryuuta/photon). Indicating that the original repository works well.
* The 1st attached file i.e. the bad one was created by modifying the repository to generate a CBDDLP/Photon file from a collection of layer/slice PNG files.

I'm trying to figure out why my modification to the repository is generating faulty files! 😕

**Reply by sn4k3** [reply]
well your preview data was incomplete/bad and layer exposure was set to a very high value. Other than that everything else look ok.
You also have this in Go: https://github.com/ezrec/uv3dp Much more modern, complete and support much more files.
Also you can use UVtools Core to generate files if you want but only with C# or powershell: https://github.com/sn4k3/UVtools/tree/master/UVtools.Powershell

UVtools GUI also allow you to import a collection of PNGs to replace a layer colection and upload previews

**Reply by Megidd** [reply]
I installed VS 2019 Community Edition and the tools [here](https://github.com/sn4k3/UVtools#develop-and-build-from-source) 😃 Opening the solution, there are two sub-projects marked as incompatible. Not sure why! 3 out of 5 sub-projects are built fine 😄 

![2021-06-10 14_21_28-UVtools - Microsoft Visual Studio](https://user-images.githubusercontent.com/17475482/121504585-5ed42a80-c9f7-11eb-8506-7d7f77a08767.png)

**Reply by Megidd** [reply]
Also, when running `build\CompileWindows.bat`, an error is thrown:

> MSBUILD : error MSB1009: Project file does not exist.
> Switch: UVtools.sln

![2021-06-10 14_15_30-UVtools - Microsoft Visual Studio](https://user-images.githubusercontent.com/17475482/121503839-ae662680-c9f6-11eb-8607-777881176368.png)

**Reply by Megidd** [reply]
When I try to use the debugger, this error is shown 😕 I'm not sure how to resolve it:

![2021-06-10 14_25_01-UVtools - Microsoft Visual Studio](https://user-images.githubusercontent.com/17475482/121505059-ba9eb380-c9f7-11eb-83fc-5fa090ce32e0.png)

**Reply by sn4k3** [reply]
> I installed VS 2019 Community Edition and the tools here 😃 Opening the solution, there are two sub-projects marked as incompatible. Not sure why! 3 out of 5 sub-projects are built fine 😄

Ignore them, they are the MSI installer, not required, you can also unload them.

> Also, when running build\CompileWindows.bat, an error is thrown:

Strange, it works for me:

```
UVtools.sln Compile
"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\MSBuild.exe"
Microsoft (R) Build Engine version 16.10.1+2fd48ab73 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 10/06/2021 14:30:16.
Project "D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Release|Any CPU".
ValidateProjects:
  The project "UVtools.InstallerMM" is not selected for building in solution configuration "Release|Any CPU".
  The project "UVtools.Installer" is not selected for building in solution configuration "Release|Any CPU".
Project "D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.sln" (1) is building "D:\Tiago\Dropbox\Programming\C#\UVtools\
UVtools.Core\UVtools.Core.csproj" (2) on node 1 (default targets).
CoreGenerateAssemblyInfo:
Skipping target "CoreGenerateAssemblyInfo" because all output files are up-to-date with respect to the input files.
CoreCompile:
```

> When I try to use the debugger, this error is shown 😕 I'm not sure how to resolve it:

You are trying to debug a library (UVtools.Core), you need to switch for UVtools.WPF.
Right click on **UVtools.WPF** on right panel, and set it as default project.
OR click on the list where says UVtools.Core near the green debug button and switch to UVtools.WPF
